lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -335,30 +393,47 @@ def __init__(self, config: PbenchServerConfig, logger: Logger, schema: Schema):
         self.es_url = f"http://{host}:{port}"
         self.schema = schema
 
-    def _get_user_term(self, user: str) -> dict:
+    def _get_user_term(self, json: Dict[str, Any]) -> dict:
         """
-        _get_user_term Generate the user term for an Elasticsearch document
-        query. If the specified user parameter is not None, search for
-        documents with that value as the authorized owner. If the user
-        parameter is None, instead search for documents with public access.
+        Generate the user term for an Elasticsearch document query. If the
+        specified "user" parameter is not None, search for documents with that
+        value as the authorized owner. If the "access" parameter is "public"
+        or if the "user" parameter is missing (or None or ""), search for
+        published data.
+
+        TODO: Currently this code supports either public or private data, but
+        not both. The syntax here allows {"user": "me", "access": "public"}
+        to combine the terms; supporting combinations requires working out
+        the right Elasticsearch binary query syntax, and I'll plan to add that
+        in a separate PR.
+
+        TODO: does {"user": "me", "access": "public"} mean we return all data
+        owned by "me" *AND* all published data, or all data owned by "me" which
+        has been published? Do we want to support both modes? This may depend
+        on how we support "publishing" data; TBD.
```

Comment:
Rather, #2370 was spawned from this TODO and from a desire to define consistent behavior and move to implement that; the implementation here doesn't try to subsume the eventual outcome of #2370. And, yes, the TODO should be updated to reference the issue and probably reworded.

Reply:
My concern is that, if we merge the approach represented by the code in this PR, then, in the interim between now and when #2370 is addressed, some client (which shall remain nameless) might come to depend on the semantics which this PR supports, and we'll have some integration pain when the changes for #2370 are merged in.

So, I'd rather see this PR implement as much of #2370 as practical, and have it _not_ implement anything which conflicts with #2370.  (And, maybe that is what it's doing...my brain has been too full of late to properly embrace all this stuff....)

Reply:
I'm probably missing something, but it looks to me like there are only three places which need to be changed, corresponding with the three calls to `_get_user_term()`.  All I _think_ you need to do is to split the access stuff out of `_get_user_term()` into a separate method, say, `_get_access_term()`, and add a call to it at those three sites.  E.g., in `DatasetsList.assemble()`, you could remove the direct call to `_get_user_term()` from the dictionary instantiation and set it up like this:
```
        ret_val = {
            "path": f"/{uri_fragment}/_search",
            "kwargs": {
                "json": {
...
                    "query": {
                        "bool": {
                            "filter": [
                                ### Remove _get_user_term() call from here ###
                                {"term": {"run.controller": controller}},
                            ]
                        }
                    },
                    "size": 5000,
                },
                "params": {"ignore_unavailable": "true"},
            },
        }

        user_term = _get_user_term(json_data)
        if user_term:
            ret_val["kwargs"]["json"]["query"]["bool"]["filter"].append({"term": user_term})

        access_term = _get_access_term(json_data)
        if access_term:
            ret_val["kwargs"]["json"]["query"]["bool"]["filter"].append({"term": access_term})

        return ret_val
```

Reply:
They're not independent. Right now it's "one or the other"; once I finish working out the Elasticsearch syntax for #2370  there'll be some conditional combination query terms probably embedded in a `"bool": { "should": [ ... ] }` and I want that logic encapsulated and isolated from the callers. Your suggestion would require that the caller understand whether they're mutually exclusive and how to combine them, and I'd rather avoid that. (Whether I can transparently patch in the necessary phrase without more awareness of the outside query remains to be seen, but I intend to try.)

Reply:
My understanding of #2370 (which, admittedly, I haven't re-read in almost a couple of weeks, now...) was not that `user` and `access` were "mutually exclusive" but rather that they were _orthogonal_.  And, I think that the above code accomplishes that:  it filters on `user`, if it is present; it filters on `access`, if it is present; it filters on _both_ if they are both present; and, I believe it also works if neither is present.  (I think you can do it all in ES "filter context" -- I don't think you need to resort to fuzzy things like "should" clauses.)

Reply:
OK, after talking with Dave, I've concluded that I was indeed missing something and my solution above would likely "leak".  Moreover, I've concluded that, in our present model, `user` and `access` are _not_ orthogonal to each other.

So, here's what I (now...) think the algorithm should look like:
```
if the request specifies a target "user", use the value in a user term in the ES query; otherwise omit it.
if the requestor is authorized and either targets itself or is an admin:
    use the "access" value in the access term if specified; otherwise omit it
else:
    the ES access term must be "public" (either force it or report an error)
```
But, Dave, if that doesn't pique your interest (e.g., because in this go 'round we're limited to a single term), we can worry about it some other time.